File: cpp/Day2.cpp
#include "Day2.h"

void Day2::part1()
{
    vector<string> lines;
    Utils::read_lines_from_file("../data/day2.txt", lines);

    int h_pos = 0, v_pos = 0;

    for (string line : lines)
    {
        vector<string> parts;
        Utils::split_string(line, " ", parts);
 
        string direction = parts[0];
        int value = stoi(parts[1]);

        if (direction == "forward")
            h_pos += value;
        else if (direction == "down")
            v_pos += value;
        else if (direction == "up")
            v_pos -= value;
        else exit(EXIT_FAILURE);
    }

    cout << "Day:  2 | Part: 1 | Result: " << h_pos * v_pos << endl;
}

void Day2::part2()
{
    vector<string> lines;
    Utils::read_lines_from_file("../data/day2.txt", lines);

    int h_pos = 0, v_pos = 0, aim = 0;

    for (string line : lines)
    {
        vector<string> parts;
        Utils::split_string(line, " ", parts);

        string direction = parts[0];
        int value = stoi(parts[1]);

        if (direction == "forward")
        {
            h_pos += value;
            v_pos += aim * value;
        }
        else if (direction == "down")
            aim += value;
        else if (direction == "up")
            aim -= value;
        else exit(EXIT_FAILURE);
    }

    cout << "Day:  2 | Part: 2 | Result: " << h_pos * v_pos << endl;
}

File: php/Day2.php
<?php
    require_once('Utils.php');

    function Day2_Part1() {
        $h_pos = 0;
        $v_pos = 0;

        foreach (read_file_lines("../data/day2.txt") as $line) {
            $value = (int) explode(' ',  $line)[1];

            switch (explode(' ', $line)[0]) {
                case "forward":
                    $h_pos += $value;
                    break;
                case "down":
                    $v_pos += $value;
                    break;
                case "up":
                    $v_pos -= $value;
                    break;
                default:
                    throw new Exception("Unsupported direction " . explode(' ', $line)[0]);
            }
        }
        
        echo "Day:  2 | Part: 1 | Result: " . ($h_pos * $v_pos) . "<br>";
    }

    function Day2_Part2() {
        $h_pos = 0;
        $v_pos = 0;
        $aim = 0;

        foreach (read_file_lines("../data/day2.txt") as $line) {
            $value = (int) explode(' ',  $line)[1];

            switch (explode(' ', $line)[0]) {
                case "forward":
                    $h_pos += $value;
                    $v_pos += $value * $aim;
                    break;
                case "down":
                    $aim += $value;
                    break;
                case "up":
                    $aim -= $value;
                    break;
                default:
                    throw new Exception("Unsupported direction " . explode(' ', $line)[0]);
            }
        }
        
        echo "Day:  2 | Part: 1 | Result: " . ($h_pos * $v_pos) . "<br>";

    }

    Day2_Part1();
    Day2_Part2();